### PR TITLE
Add guest blog post by Seb James on data viz with C++

### PR DESCRIPTION
**Pull request by willfurnass**
@sebjameswml - happy for this to appear here and does it look okay to you?

### CODE REVIEW

File: _posts/2021-01-25-datavis-cpp.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+So why this post about data visualisation in C++? While the wider data-vis community won’t see C++ as a primary tool, there is a use case for researchers who write numerically intensive models - any model which 'contains' a lot of numbers and benefits from making maximal use of available computational resources. My work falls into this category. Although I don't need racks of servers to run my 2D reaction-diffusion models, each simulation takes a minute or two to run, making use of multiple CPU cores. On past experience, I'd expect these programs to run about an order of magnitude slower in Python or MATLAB. In practice, rather than waiting longer, I would limit the resolution, or scope of the model to keep the time it required to run at an optimum value. A batch of simulations running overnight is much more useful than a batch which needs a week to complete. However, by limiting the scope of the model, I might be ignoring science that would only present itself in a higher resolution system.  Put simply, using C++ allows me to get the most out of the computer hardware I have access to.
```
Comment (Robadob) [inline _posts/2021-01-25-datavis-cpp.md]: There's a double space before the final sentence (and I can't see double space after fullstop being used elsewhere.)

File: _posts/2021-01-25-datavis-cpp.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+To write the code, I was inspired by gaming technology. My kids play a game called Garry's Mod. Garry's Mod isn't actually a game at all, it's a physics sandbox; a 3D environment into which you can place models with which you can interact. Really, it's an example of a game engine. I imagined building a similar world into which I could place graph models which would update as my simulations progressed. I wanted to fly around the graphs, viewing them from any angle and bringing as many into the view as necessary. A lot of the work rendering the scene would be carried out on my graphics processor, leaving my CPU free to compute the model. The output of the visualisation, which could be captured as screenshots, wouldn't necessarily need to be of publication quality - a traditional approach using Python and matplotlib could be used to generate additional analysis and final graphs for a paper.
```
Comment (Robadob) [inline _posts/2021-01-25-datavis-cpp.md:35]: The sequel to gmod is under development now! I played alot of gmod about a decade ago.
https://sbox.facepunch.com/news

File: _posts/2021-01-25-datavis-cpp.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+I chose an architecture in which each graph visualisation was a class derived from a base class called [morph::VisualModel](https://github.com/ABRG-Models/morphologica/blob/main/morph/VisualModel.h). The first VisualModels were [HexGridVisual](https://github.com/ABRG-Models/morphologica/blob/main/morph/HexGridVisual.h) and [CoordArrows](https://github.com/ABRG-Models/morphologica/blob/main/morph/CoordArrows.h); these were the only objects in my first scenes. Later, I added other VisualModels, including morph::GraphVisual, which makes it easy to create a regular 2D line or scatter graph. This has given me the opportunity to try to develop a more intuitive API than the one provided by matplotlib in Python (and copied in matplot++), which I have always found confusing (there always seem to be too many ways to achieve a result). In morph::Visual, I can make decisions about how I think the API should work and I can favour simplicity over flexibility.
```
Comment (Robadob) [inline _posts/2021-01-25-datavis-cpp.md]: if morph::GraphVisual, morph::Visual refer to code things they might read better as `morph::GraphVisual` and `morph::Visual` respectively. Possible other cases of this elsewhere (I can see one in the previous paragraph too).

File: _posts/2021-01-25-datavis-cpp.md (hunk 1 above)
Comment (willfurnass) [reply _posts/2021-01-25-datavis-cpp.md]: Ta.  Going to start calling you 'hawk-eye'.

File: _posts/2021-01-25-datavis-cpp.md (hunk 3 above)
Comment (willfurnass) [reply _posts/2021-01-25-datavis-cpp.md]: Good suggestion; done in new commit.

**Comment by Robadob** [comment]
The blog post reads fine, although I'm not sure how wide the audience is (as your blogpost admits, most people do everything with python/R).

----------------------
*Unrelated comments*

Personally for C++ rendering, during my PhD alongside @ptheywood, I developed [sdl_exp](https://github.com/Robadob/sdl_exp) as a very general low-level wrapper over OpenGL (and personal experimentation with OpenGL). We work with complex simulations, with potentially millions of agents, so we've mostly used hard-fork's of it for real-time visual model validation (e.g. rendering agents as coloured spheres/people). In particular, as we use CUDA for our simulations it's capable of rendering data directly on the GPU, rather than copying it to the GPU each frame.

With regards to real-time graphs, the models we use are highly stochastic so before graphing we tend to collect timestep data from hundreds of runs, to produce graphs of average + standard deviation (using Python). The current project I'm attached to, I'm actually doing this to validate against identical runs in a much slower Python version of the model.

I've had a look through the library and it seems thorough and well-written, you've even used Freetype for the text (I've done this, and the character placement was alot of faff, I think I still have bugs around newline characters). I did notice however that you've written your own graphics math primitives (e.g. matrix3x3), [GLM](https://github.com/g-truc/glm) might be of interest to you, its a header library that aims to provide a C++ interface to all the maths functionality available inside GLSL shaders (and more).

@mondus has said he hasn't got time to review this currently, but to me it seems like something he might be interested in. I'm also sure he'd also be interested to know whether the scale of your models could be increased further with GPUs (e.g. are there highly parallel operations).

**Review by Robadob — APPROVED** [review]
See comment/s

**Pull request merged**